Adding Canary Release EN (#267)
oam-dev/kubevela.io: docs/case-studies/canary-blue-green.md

---
title:  Canary Release
---

> This article will introduce that based on [Istio](https://istio.io/latest/) how KubeVela implement classic microservice scenarios [bookinfo](https://istio.io/latest/docs/examples/bookinfo/?ie=utf-8&hl=en&docs-search=Canary) as Canary Release.

Reviewer: ```suggestion
This article will introduce that based on [Istio](https://istio.io/latest/) how KubeVela implement classic microservice scenarios [bookinfo](https://istio.io/latest/docs/examples/bookinfo/?ie=utf-8&hl=en&docs-search=Canary) as Canary Release.
```
```suggestion
This article introduces how KubeVela integrates [Istio](https://istio.io/latest/) to do a canary release.
```

## Praparation

> Open the Istio cluster plugin.

Reviewer: ```suggestion
Install the Istio cluster plugin.
```

```shell
vela addon enable istio
```

> Wait for a while and confirm if the status of the cluster plugin is `success`, indicating that it is ready.

Reviewer: ```suggestion
Wait until the status of the cluster plugin is `success`, which indicates it is ready.
```

```shell

> kubectl get initializer -n istio-system istio

Reviewer: initializer 不是要deprecate 吗？这个怎么打算？

Author: 删掉了。不应该透出 initializer。后续 addon 会补上对是否成功的检查。

NAME    PHASE          AGE
istio   success       4h47m
```

> Because the following example runs in the default namespace, the default namespace needs to be labeled with Istio auto-injecting sidecar.

Reviewer: ```suggestion
The default namespace needs to be labeled so that Istio will auto-inject sidecar.
```

```shell
kubectl label namespace default istio-injection=enabled
```

## Initial deployment

> Execute the following command to deploy the Application of `bookinfo`.

Reviewer: ```suggestion
Deploy the Application of `bookinfo`:
```

```shell
kubectl apply -f https://github.com/oam-dev/kubevela/blob/master/docs/examples/canary-rollout-use-case/first-deploy.yaml
```

The component architecture and  relationship of the application are as follows:

![book-info-struct](../resources/book-info-struct.jpg)

> This Application contains four Components, each of which is configured with a Trait of `expose` to expose services in the cluster.

Reviewer: ```suggestion
This Application has four Components, each configured with an`expose` Trait to expose cluster-level service.
```

> The component of `productpage` is also configured with an istio-gateway Trait, allowing the Component to receive traffic entering the cluster. This Trait is achieved by setting `gateway:ingressgateway` to use Istio's default gateway, and setting `hosts: "*"` to specify that requests carrying any host information can enter the gateway.

Reviewer: ```suggestion
The `productpage` component is also configured with an `istio-gateway` Trait, allowing the Component to receive traffic coming from outside the cluster. The example below show that it sets `gateway:ingressgateway` to use Istio's default gateway, and `hosts: "*"` to specify that any request can enter the gateway.
```

```shell
...
    - name: productpage
      type: webservice
      properties:
          image: docker.io/istio/examples-bookinfo-productpage-v1:1.16.2
          port: 9080

      traits:
          - type: expose
            properties:
              port:
                - 9080

          - type: istio-gateway
            properties:
              hosts:
                - "*"
              gateway: ingressgateway
              match:
                - exact: /productpage
                - prefix: /static
                - exact: /login
                - prefix: /api/v1/products
              port: 9080
...
```

> You can map the port of the gateway to the local by executing the following command.

Reviewer: ```suggestion
You can port-forward to the gateway as follows:
```

```shell
kubectl port-forward service/istio-ingressgateway -n istio-system 19082:80
```
Visit `127.0.0.1:19082` through the browser and you will see the following page.

![pic-v2](../resources/canary-pic-v2.jpg)

## Canary Release

Next, we take the `reviews` Component as an example to simulate the complete process of a canary release, and first upgrade a part of the component instances, and adjust the traffic at the same time, so as to achieve the purpose of progressive canary release.

Execute the following command to update the application.
```shell
kubectl apply -f https://github.com/oam-dev/kubevela/blob/master/docs/examples/canary-rollout-use-case/rollout-v2.yaml
```

> This operation updated the mirror of the `reviews` Component from the previous v2 to v3. At the same time, the Rollout Trait of the `reviews` Component specifies that the number of target instances to be upgraded is two, which are upgraded in two batches, with one instance in each batch.

Reviewer: ```suggestion
This operation updates the mirror of the `reviews` Component from the previous v2 to v3. At the same time, the Rollout Trait of the `reviews` Component specifies that the number of target instances to be upgraded is two, which are upgraded in two batches, with one instance in each batch.
```

```

This update deletes the previously defined workflow to terminate the execution of the workflow.

And by modifying the `targetRevision` of the Rollout Trait to point to the previous component version `reviews-v1`. In addition, this update also removes the canary-traffic Trait of the Component, and puts all traffic on the same component version `reviews-v1`.

```shell
...
    - name: reviews
      type: webservice
      properties:
        image: docker.io/istio/examples-bookinfo-reviews-v3:1.16.2
        port: 9080
        volumes:
          - name: wlp-output
            type: emptyDir
            mountPath: /opt/ibm/wlp/output
          - name: tmp
            type: emptyDir
            mountPath: /tmp

      traits:
        - type: expose
          properties:
            port:
              - 9080

        - type: rollout
          properties:
            targetRevision: reviews-v1
            batchPartition: 1
            targetSize: 2
            # This means to rollout two more replicas in two batches.
            rolloutBatches:
              - replicas: 2
...
```

> Continue to visit the website on the browser, you will find that the five-pointed star has changed back to black.

Reviewer: 建议贴图。
